File: sdks/python/src/akasha/http_client.py
"""
Akasha HTTP Client — Lightweight client using the REST API.

For environments where gRPC is not available or when simplicity is preferred.
Uses httpx for both sync and async HTTP support.
"""

from __future__ import annotations

import json
import logging
from typing import Any, AsyncIterator, Optional

import httpx

from akasha.models import AkashaEvent, Record

logger = logging.getLogger("akasha.http_client")


class AkashaHttpClient:
    """
    HTTP/REST client for Akasha — lightweight alternative to gRPC.

    Usage:
        client = AkashaHttpClient("http://localhost:7777")

        record = client.put("agents/my-agent/state", {"status": "thinking"})
        record = client.get("agents/my-agent/state")
        records = client.query("agents/*/state")
        agents = client.list_agents()
        tree = client.tree()

        client.close()
    """

    def __init__(
        self,
        base_url: str = "http://localhost:7777",
        *,
        timeout: float = 10.0,
        max_retries: int = 3,
    ):
        self._base_url = base_url.rstrip("/")
        self._client = httpx.Client(
            base_url=self._base_url,
            timeout=timeout,
            transport=httpx.HTTPTransport(retries=max_retries),
        )
        logger.info(f"Akasha HTTP client connected to {base_url}")

    def close(self):
        """Close the HTTP client."""
        self._client.close()

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()

    # -- Core CRUD --

    def put(
        self,
        path: str,
        value: Any,
        *,
        ttl_seconds: Optional[float] = None,
        tags: Optional[dict[str, str]] = None,
    ) -> Record:
        """Write a record via REST API."""
        body: dict[str, Any] = {"value": value}
        if ttl_seconds is not None:
            body["ttl_seconds"] = ttl_seconds
        if tags:
            body["tags"] = tags

        response = self._client.post(
            f"/api/v1/records/{path}",
            json=body,
        )
        response.raise_for_status()
        return Record.from_dict(response.json())

    def get(self, path: str) -> Optional[Record]:
        """Read a record by path."""
        response = self._client.get(f"/api/v1/records/{path}")
        if response.status_code == 404:
            return None
        response.raise_for_status()
        return Record.from_dict(response.json())

    def delete(self, path: str) -> bool:
        """Delete a record."""
        response = self._client.delete(f"/api/v1/records/{path}")
        return response.status_code == 204

    # -- Queries --

    def query(
        self,
        pattern: str,
        *,
        limit: Optional[int] = None,
    ) -> list[Record]:
        """Query records matching a glob pattern."""
        params: dict[str, Any] = {"pattern": pattern}
        if limit is not None:
            params["limit"] = limit

        response = self._client.get("/api/v1/query", params=params)
        response.raise_for_status()
        return [Record.from_dict(r) for r in response.json()]

    # -- Convenience --

    def list_agents(self) -> list[dict]:
        """List all registered agents."""
        response = self._client.get("/api/v1/agents")
        response.raise_for_status()
        return response.json()

    def tree(self) -> dict[str, Any]:
        """Get the full state tree snapshot."""
        response = self._client.get("/api/v1/tree")
        response.raise_for_status()
        return response.json()

    def health(self) -> dict:
        """Health check."""
        response = self._client.get("/api/v1/health")
        response.raise_for_status()
        return response.json()

    def metrics(self) -> dict:
        """Get server metrics."""
        response = self._client.get("/api/v1/metrics")
        response.raise_for_status()
        return response.json()


class AsyncAkashaHttpClient:
    """
    Async HTTP client for Akasha — for asyncio environments without gRPC.

    Usage:
        async with AsyncAkashaHttpClient("http://localhost:7777") as client:
            await client.put("agents/my-agent/state", {"status": "thinking"})
    """

    def __init__(
        self,
        base_url: str = "http://localhost:7777",
        *,
        timeout: float = 10.0,
        max_retries: int = 3,
    ):
        self._base_url = base_url.rstrip("/")
        self._client = httpx.AsyncClient(
            base_url=self._base_url,
            timeout=timeout,
            transport=httpx.AsyncHTTPTransport(retries=max_retries),
        )

    async def close(self):
        await self._client.aclose()

    async def __aenter__(self):
        return self

    async def __aexit__(self, *args):
        await self.close()

    async def put(
        self,
        path: str,
        value: Any,
        *,
        ttl_seconds: Optional[float] = None,
        tags: Optional[dict[str, str]] = None,
    ) -> Record:
        body: dict[str, Any] = {"value": value}
        if ttl_seconds is not None:
            body["ttl_seconds"] = ttl_seconds
        if tags:
            body["tags"] = tags

        response = await self._client.post(f"/api/v1/records/{path}", json=body)
        response.raise_for_status()
        return Record.from_dict(response.json())

    async def get(self, path: str) -> Optional[Record]:
        response = await self._client.get(f"/api/v1/records/{path}")
        if response.status_code == 404:
            return None
        response.raise_for_status()
        return Record.from_dict(response.json())

    async def delete(self, path: str) -> bool:
        response = await self._client.delete(f"/api/v1/records/{path}")
        return response.status_code == 204

    async def query(self, pattern: str, *, limit: Optional[int] = None) -> list[Record]:
        params: dict[str, Any] = {"pattern": pattern}
        if limit is not None:
            params["limit"] = limit
        response = await self._client.get("/api/v1/query", params=params)
        response.raise_for_status()
        return [Record.from_dict(r) for r in response.json()]

    async def list_agents(self) -> list[dict]:
        response = await self._client.get("/api/v1/agents")
        response.raise_for_status()
        return response.json()

    async def tree(self) -> dict[str, Any]:
        response = await self._client.get("/api/v1/tree")
        response.raise_for_status()
        return response.json()

    async def health(self) -> dict:
        response = await self._client.get("/api/v1/health")
        response.raise_for_status()
        return response.json()

    async def metrics(self) -> dict:
        response = await self._client.get("/api/v1/metrics")
        response.raise_for_status()
        return response.json()

File: sdks/node/src/http-client.ts
/**
 * Akasha HTTP Client — Lightweight REST client for Node.js.
 *
 * Uses the native fetch API (Node 18+) with no extra dependencies.
 * Great for scripts, dashboards, and environments where gRPC is overkill.
 */

import { AkashaRecord, AkashaEvent, recordFromJson, eventFromJson } from './models';

export interface AkashaHttpClientOptions {
  /** HTTP base URL (default: "http://localhost:7777") */
  baseUrl?: string;
  /** Default timeout in milliseconds (default: 10000) */
  timeout?: number;
}

export class AkashaHttpClient {
  private baseUrl: string;
  private timeout: number;

  constructor(options: AkashaHttpClientOptions = {}) {
    this.baseUrl = (options.baseUrl ?? 'http://localhost:7777').replace(/\/$/, '');
    this.timeout = options.timeout ?? 10000;
  }

  // -- Core CRUD --

  /** Write a record via REST API. */
  async put(
    path: string,
    value: unknown,
    options: { ttlSeconds?: number; tags?: { [key: string]: string } } = {}
  ): Promise<AkashaRecord> {
    const body: any = { value };
    if (options.ttlSeconds !== undefined) body.ttl_seconds = options.ttlSeconds;
    if (options.tags) body.tags = options.tags;

    const res = await this.fetch(`/api/v1/records/${path}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    });

    return recordFromJson(await res.json());
  }

  /** Read a record by path. Returns null if not found. */
  async get(path: string): Promise<AkashaRecord | null> {
    const res = await this.fetch(`/api/v1/records/${path}`);
    if (res.status === 404) return null;
    return recordFromJson(await res.json());
  }

  /** Delete a record. */
  async delete(path: string): Promise<boolean> {
    const res = await this.fetch(`/api/v1/records/${path}`, { method: 'DELETE' });
    return res.status === 204;
  }

  // -- Queries --

  /** Query records matching a glob pattern. */
  async query(
    pattern: string,
    options: { limit?: number } = {}
  ): Promise<AkashaRecord[]> {
    const params = new URLSearchParams({ pattern });
    if (options.limit !== undefined) params.set('limit', String(options.limit));

    const res = await this.fetch(`/api/v1/query?${params}`);
    const data = await res.json();
    return (data as any[]).map(recordFromJson);
  }

  // -- Convenience --

  /** List all registered agents. */
  async listAgents(): Promise<any[]> {
    const res = await this.fetch('/api/v1/agents');
    return res.json();
  }

  /** Get the full state tree snapshot. */
  async tree(): Promise<{ [key: string]: unknown }> {
    const res = await this.fetch('/api/v1/tree');
    return res.json();
  }

  /** Health check. */
  async health(): Promise<{ status: string; name: string; version: string; records: number }> {
    const res = await this.fetch('/api/v1/health');
    return res.json();
  }

  /** Get server metrics. */
  async metrics(): Promise<any> {
    const res = await this.fetch('/api/v1/metrics');
    return res.json();
  }

  // -- Stigmergy (Pheromone System) --

  /**
   * Deposit a pheromone trace. If a pheromone already exists at the
   * same trail, it is reinforced (intensities sum).
   */
  async depositPheromone(
    trail: string,
    options: {
      signalType?: string;
      emitter?: string;
      intensity?: number;
      halfLifeSecs?: number;
      payload?: unknown;
    } = {}
  ): Promise<any> {
    const body = {
      trail,
      signal_type: options.signalType ?? 'success',
      emitter: options.emitter ?? 'anonymous',
      intensity: options.intensity ?? 1.0,
      half_life_secs: options.halfLifeSecs ?? 3600,
      payload: options.payload,
    };

    const res = await this.fetch('/api/v1/pheromones', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    });

    return res.json();
  }

  /** Sense active pheromones. */
  async sensePheromones(): Promise<any[]> {
    const res = await this.fetch('/api/v1/pheromones');
    return res.json();
  }

  // -- Cognitive Fabric --

  /**
   * Write a record to a specific memory layer.
   * @param layer One of: "working", "episodic", "semantic", "procedural"
   */
  async writeMemory(
    layer: string,
    namespace: string,
    key: string,
    value: unknown,
    options: { ttlSeconds?: number; tags?: { [key: string]: string } } = {}
  ): Promise<AkashaRecord> {
    const path = `memory/${layer}/${namespace}/${key}`;
    const merged = { ...options.tags, _memory_layer: layer };
    return this.put(path, value, { ttlSeconds: options.ttlSeconds, tags: merged });
  }

  /** Read a record from a memory layer. */
  async readMemory(
    layer: string,
    namespace: string,
    key: string
  ): Promise<AkashaRecord | null> {
    const path = `memory/${layer}/${namespace}/${key}`;
    return this.get(path);
  }

  /** Get Nidra consolidation status. */
  async nidraStatus(): Promise<any> {
    const res = await this.fetch('/api/v1/nidra/status');
    return res.json();
  }

  /** Get record counts per memory layer. */
  async memoryLayers(): Promise<any> {
    const res = await this.fetch('/api/v1/memory/layers');
    return res.json();
  }

  // -- WebSocket Subscription --

  /**
   * Subscribe to real-time events via WebSocket.
   * Requires the 'ws' package or a WebSocket-compatible environment.
   *
   * @param pattern Glob pattern to filter events
   * @param callback Function called for each event
   * @returns A close function to stop the subscription
   */
  subscribeWs(
    pattern: string,
    callback: (event: AkashaEvent) => void
  ): { close: () => void } {
    const wsUrl = this.baseUrl.replace(/^http/, 'ws');
    const ws = new WebSocket(`${wsUrl}/api/v1/stream?pattern=${encodeURIComponent(pattern)}`);

    ws.onmessage = (msg) => {
      try {
        const data = JSON.parse(typeof msg.data === 'string' ? msg.data : msg.data.toString());
        callback(eventFromJson(data));
      } catch (err) {
        console.warn('[Akasha] Failed to parse WebSocket event:', err);
      }
    };

    ws.onerror = (err) => {
      console.error('[Akasha] WebSocket error:', err);
    };

    return {
      close: () => ws.close(),
    };
  }

  // -- Internal --

  private async fetch(path: string, init?: RequestInit): Promise<Response> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);

    try {
      const res = await globalThis.fetch(`${this.baseUrl}${path}`, {
        ...init,
        signal: controller.signal,
      });

      if (!res.ok && res.status !== 404 && res.status !== 204) {
        const body = await res.text().catch(() => '');
        throw new Error(`Akasha HTTP ${res.status}: ${body}`);
      }

      return res;
    } finally {
      clearTimeout(timer);
    }
  }
}
